File: officialdrops/sheets_append_row.ts
/**
 * sheets_append_row — official scripted connector (replaces the native Go drop).
 * Appends rows (from the 'rows' input) to a Google Sheet tab.
 */

const SHEETS_API_BASE = "https://sheets.googleapis.com/v4";

export default {
  manifest: {
    id: "sheets_append_row",
    version: "2.0.0",
    label: "Sheets append rows",
    summary: "Append rows to a Google Sheet tab, with the first row picking column order from headers.",
    description:
      "Append rows to a Google Sheet. Rows come in on the 'rows' input; column order is the optional 'headers' input or derived (sorted) from row keys.",
    integration: "Google Sheets",
    category: "network",
    icon: "file-output",
    brandLogo: "/brands/sheets.svg",
    color: "#0F9D58",
    tags: ["sheets", "google", "append", "log", "etl"],
    requiresConnections: [{ kind: "oauth", name: "google", note: "Google OAuth — sheets/drive scope." }],
    inputs: [
      { port: "rows", label: "Rows", required: true, mime: ["application/json"] },
      { port: "headers", label: "Headers (column order)", mime: ["application/json"] },
    ],
    outputs: [{ port: "meta", label: "Append metadata", mime: ["application/json"] }],
    idempotent: false,
    retryPolicy: "exponential_backoff",
    paramsSchema: {
      type: "object",
      properties: {
        base_url: { type: "string", description: "Override the API host (proxy / self-hosted / testing)." },
        account: { type: "string", default: "default" },
        token: { type: "string", description: "Raw access token; overrides 'account'." },
        spreadsheet_id: { type: "string", description: "Sheet ID or full URL." },
        range: { type: "string", default: "Sheet1", description: "Sheet name or range; appends after the last populated row." },
        value_input_option: { type: "string", enum: ["USER_ENTERED", "RAW"], default: "USER_ENTERED" },
        insert_data_option: { type: "string", enum: ["INSERT_ROWS", "OVERWRITE"], default: "INSERT_ROWS" },
        timeout_ms: { type: "integer", default: 15000, minimum: 1 },
      },
      required: ["spreadsheet_id"],
    },
    examples: [
      { title: "Append to the default sheet", params: { account: "default", spreadsheet_id: "1AbcDE...", range: "Sheet1" }, notes: "Rows come in on the 'rows' input port." },
    ],
  },

  async run(ctx: any) {
    const p = ctx.params || {};
    const base = String(p.base_url || SHEETS_API_BASE).replace(/\/+$/, "");
    const id = sheetId(p.spreadsheet_id);
    if (!id) throw new DropError("bad_param", "'spreadsheet_id' is required");
    const token = await googleToken(ctx, p);

    if (!ctx.inputs.has("rows")) throw new DropError("missing_input", "input port 'rows' is required");
    const rowsVal = ctx.inputs.get("rows");
    if (!Array.isArray(rowsVal)) throw new DropError("bad_input", "'rows' must be an array of objects");
    const rows: Record<string, any>[] = rowsVal;

    let headers: string[];
    const headersVal = ctx.inputs.has("headers") ? ctx.inputs.get("headers") : undefined;
    if (Array.isArray(headersVal)) headers = headersVal.map(String);
    else headers = deriveHeaders(rows);

    const values = rows.map((row) => headers.map((h) => (h in (row || {}) ? row[h] : "")));
    const range = String(p.range || "Sheet1");

    if (values.length === 0) {
      return { meta: { appended_rows: 0 } };
    }

    const res = await ctx.fetch(`${base}/spreadsheets/${encodeURIComponent(id)}/values/${encodeURIComponent(range)}:append`, {
      method: "POST",
      query: {
        valueInputOption: String(p.value_input_option || "USER_ENTERED"),
        insertDataOption: String(p.insert_data_option || "INSERT_ROWS"),
      },
      headers: { Authorization: `Bearer ${token}`, "Content-Type": "application/json; charset=utf-8" },
      body: JSON.stringify({ range, majorDimension: "ROWS", values }),
      timeoutMs: Number(p.timeout_ms) || 15000,
    });
    if (!res.ok) throw new DropError("sheets_error", `Sheets returned ${res.status}: ${sheetsErr(await res.text())}`);
    const parsed: any = await res.json();
    const u = parsed.updates || {};
    return {
      meta: {
        appended_rows: values.length,
        spreadsheet_id: id,
        updated_range: u.updatedRange || "",
        updated_rows: u.updatedRows || 0,
        updated_columns: u.updatedColumns || 0,
        updated_cells: u.updatedCells || 0,
      },
    };
  },
};

// deriveHeaders is the union of row keys, sorted (matches the SQL drops' rule).
function deriveHeaders(rows: Record<string, any>[]): string[] {
  const seen: Record<string, boolean> = {};
  for (const r of rows) for (const k of Object.keys(r || {})) seen[k] = true;
  return Object.keys(seen).sort();
}

function sheetId(raw: any): string {
  const s = String(raw || "").trim();
  const m = s.match(/\/d\/([a-zA-Z0-9-_]+)/);
  return m ? m[1] : s;
}

function sheetsErr(text: string): string {
  try {
    const e = JSON.parse(text);
    if (e && e.error && e.error.message) return e.error.message;
  } catch (_e) {
    // not JSON
  }
  return text.slice(0, 512);
}

async function googleToken(ctx: any, p: any): Promise<string> {
  const t = String(p.token || "").trim();
  if (t) return t;
  if (!ctx.auth) throw new DropError("auth", "no token supplied and OAuth is not configured");
  return ctx.auth.token("google", p.account || "default");
}

File: integrations/sheets/sheets_append_row.go
package sheets

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"time"

	"git.sr.ht/~klahr/hazy-flow/core"
	"git.sr.ht/~klahr/hazy-flow/engine"
	"git.sr.ht/~klahr/hazy-flow/integrations/internal/params"
)

func init() {
	engine.Register(engine.NativeDrop{
		Manifest: core.Manifest{
			ID:             "sheets_append_row",
			Version:        "1.0",
			Label:          "Sheets append rows",
			Color:          "#0F9D58",
			Icon:           "file-output",
			BrandLogo:      "/brands/sheets.svg",
			Category:       "network",
			Provider:       "internal",
			Integration:    "Google Sheets",
			Tags:           []string{"sheets", "google", "append", "log", "etl"},
			Description:    "Append rows to a Google Sheet. Useful for logging events into a spreadsheet teammates can inspect, or for keeping a Sheet in sync with another source. By default Sheets parses '30' as a number and '=SUM(A:A)' as a formula — switch to RAW mode if you want everything stored as literal strings.",
			Summary:        "Append rows to a Google Sheet tab, with the first row picking column order from headers.",
			Examples: []core.ParamsExample{
				{
					Title:  "Append to the default sheet",
					Params: json.RawMessage(`{"account":"default","spreadsheet_id":"1AbcDEFghIJklmNOPqrsTUVwxyZ_0123456789abcd","range":"Sheet1"}`),
					Notes:  "Rows come in on the 'rows' input port; the column order is derived from row keys (alphabetical) when no headers input is wired.",
				},
				{
					Title:  "Append to a named tab with RAW value mode",
					Params: json.RawMessage(`{"account":"default","spreadsheet_id":"1AbcDEFghIJklmNOPqrsTUVwxyZ_0123456789abcd","range":"event_log!A:E","value_input_option":"RAW","insert_data_option":"INSERT_ROWS"}`),
					Notes:  "RAW preserves leading zeros and stops Sheets from interpreting '=SUM(...)' as a formula.",
				},
				{
					Title:  "Append using a raw access token",
					Params: json.RawMessage(`{"token":"${secret:SHEETS_OAUTH}","spreadsheet_id":"1AbcDEFghIJklmNOPqrsTUVwxyZ_0123456789abcd","range":"Sheet1","timeout_ms":20000}`),
				},
			},
			RequiresConnections: []core.ConnectionRequirement{
				{Kind: "oauth", Name: "sheets", Note: "Google Sheets OAuth — drive.file scope."},
			},
			ExecutionModel: core.ExecutionBatch,
			ProcessModel:   core.ProcessLongLived,
			Inputs: []core.Port{
				{Port: "rows", Label: "Rows", Required: true, MIME: []string{"application/json"}},
				{Port: "headers", Label: "Headers (column order)", Required: false, MIME: []string{"application/json"}},
			},
			Outputs: []core.Port{
				{Port: "meta", Label: "Append metadata", MIME: []string{"application/json"}},
			},
			ParamsSchema: json.RawMessage(`{
				"type":"object",
				"properties":{
					"account":            {"type":"string","default":"default"},
					"token":              {"type":"string","description":"Raw access token; overrides 'account'."},
					"spreadsheet_id":     {"type":"string","description":"Either the Sheet ID (the long string between /d/ and /edit in the URL) or paste the whole URL — the drop extracts the ID for you."},
					"range":              {"type":"string","default":"Sheet1","description":"Sheet name (e.g. \"Sheet1\") or a full range (\"Sheet1!A1:Z\"). Sheets appends after the last populated row in this range."},
					"value_input_option": {"type":"string","enum":["USER_ENTERED","RAW"],"default":"USER_ENTERED","description":"USER_ENTERED parses strings as numbers/dates/formulas; RAW keeps them literal."},
					"insert_data_option": {"type":"string","enum":["INSERT_ROWS","OVERWRITE"],"default":"INSERT_ROWS","description":"INSERT_ROWS pushes existing rows down; OVERWRITE fills empty cells in place."},
					"timeout_ms":         {"type":"integer","default":15000,"minimum":1}
				},
				"required":["spreadsheet_id"]
			}`),
			Idempotent:  false, // Repeating an append duplicates rows.
			RetryPolicy: core.RetryExponentialBackoff,
		},
		Execute: executeSheetsAppendRow,
	})
}

// executeSheetsAppendRow converts the canonical {column: value} rows
// shape into Google's `[[value, ...], ...]` lists-of-lists format,
// then POSTs to the values.append endpoint. Column order comes from
// the optional `headers` input or, when absent, is derived from row
// keys (sorted alphabetically — same rule as the SQL drops).
//
// Cell value handling: pass Go-native values through to JSON. Sheets
// with value_input_option=USER_ENTERED interprets strings (parsing
// dates, numbers, formulas); ints/floats/bools become typed cells
// directly. The result matches what a human typing the value would
// see — surprising in a few edge cases (a string "01234" becomes the
// number 1234, dropping the leading zero) but the right default for
// ETL into a human-edited sheet. Pass value_input_option=RAW to
// preserve literal strings.
func executeSheetsAppendRow(ctx context.Context, job core.Job, _ chan<- core.Progress) (core.Result, error) {
	spreadsheetIDRaw, err := params.String(job.Params, "spreadsheet_id")
	if err != nil {
		return params.Err(job, "bad_param", err.Error()), nil
	}
	spreadsheetID := normalizeSpreadsheetID(spreadsheetIDRaw)
	token, err := resolveToken(ctx, job)
	if err != nil {
		return params.Err(job, "auth", err.Error()), nil
	}

	rowsRef, ok := job.Input["rows"]
	if !ok {
		return params.Err(job, "missing_input", "input port 'rows' is required"), nil
	}
	rows, err := normalizeRows(rowsRef.Inline)
	if err != nil {
		return params.Err(job, "bad_input", err.Error()), nil
	}

	var headers []string
	if h, ok := job.Input["headers"]; ok && h.Inline != nil {
		headers, err = normalizeHeaders(h.Inline)
		if err != nil {
			return params.Err(job, "bad_input", err.Error()), nil
		}
	}
	if headers == nil {
		headers = deriveHeaders(rows)
	}

	// Build the values matrix. Sheets wants [][]any with the outer
	// slice being rows and the inner being cells in column order.
	values := make([][]any, 0, len(rows))
	for _, row := range rows {
		cells := make([]any, len(headers))
		for i, h := range headers {
			if v, ok := row[h]; ok {
				cells[i] = v
			} else {
				cells[i] = ""
			}
		}
		values = append(values, cells)
	}

	if len(values) == 0 {
		// Sheets returns 400 for an empty values list. Short-circuit
		// with a clean OK + zero count rather than confusing the user
		// with a 400 about an empty operation they probably intended.
		return core.Result{
			JobID:  job.ID,
			Status: core.StatusOK,
			Output: map[string]core.Ref{
				"meta": {MIME: "application/json", Inline: map[string]any{
					"appended_rows": 0,
				}},
			},
		}, nil
	}

	rangeRef := params.StringDefault(job.Params, "range", "Sheet1")
	payload := map[string]any{
		"range":          rangeRef,
		"majorDimension": "ROWS",
		"values":         values,
	}
	body, _ := json.Marshal(payload)

	q := url.Values{}
	q.Set("valueInputOption", params.StringDefault(job.Params, "value_input_option", "USER_ENTERED"))
	q.Set("insertDataOption", params.StringDefault(job.Params, "insert_data_option", "INSERT_ROWS"))

	endpoint := fmt.Sprintf("%s/spreadsheets/%s/values/%s:append?%s",
		currentHTTPBase(),
		url.PathEscape(spreadsheetID),
		url.PathEscape(rangeRef),
		q.Encode())
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, endpoint, bytes.NewReader(body))
	if err != nil {
		return params.Err(job, "internal", err.Error()), nil
	}
	req.Header.Set("Authorization", "Bearer "+token)
	req.Header.Set("Content-Type", "application/json; charset=utf-8")

	timeoutMs := params.IntDefault(job.Params, "timeout_ms", 15000)
	client := &http.Client{Timeout: time.Duration(timeoutMs) * time.Millisecond}
	resp, err := client.Do(req)
	if err != nil {
		return params.Err(job, "send_failed", err.Error()), nil
	}
	defer resp.Body.Close()
	respBody, _ := io.ReadAll(io.LimitReader(resp.Body, 256*1024))

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return params.Err(job, "sheets_error",
			fmt.Sprintf("Sheets returned %d: %s", resp.StatusCode, extractSheetsError(respBody))), nil
	}

	var parsed map[string]any
	_ = json.Unmarshal(respBody, &parsed)
	// Sheets' append response wraps update info under `updates`.
	// updatedRange tells you exactly where the rows landed
	// ("Sheet1!A12:C13") — useful for follow-up writes to the same
	// rows or for a Slack notification ("appended to row 12-13").
	updates, _ := parsed["updates"].(map[string]any)
	meta := map[string]any{
		"appended_rows":   len(values),
		"spreadsheet_id":  spreadsheetID,
		"updated_range":   stringField(updates, "updatedRange"),
		"updated_rows":    intField(updates, "updatedRows"),
		"updated_columns": intField(updates, "updatedColumns"),
		"updated_cells":   intField(updates, "updatedCells"),
	}
	return core.Result{
		JobID:  job.ID,
		Status: core.StatusOK,
		Output: map[string]core.Ref{
			"meta": {MIME: "application/json", Inline: meta},
		},
	}, nil
}

func stringField(m map[string]any, k string) string {
	if m == nil {
		return ""
	}
	if s, ok := m[k].(string); ok {
		return s
	}
	return ""
}

func intField(m map[string]any, k string) int {
	if m == nil {
		return 0
	}
	switch v := m[k].(type) {
	case int:
		return v
	case int64:
		return int(v)
	case float64:
		return int(v)
	}
	return 0
}
